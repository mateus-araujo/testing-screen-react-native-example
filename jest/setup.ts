import 'react-native-gesture-handler/jestSetup';

require("jest-fetch-mock").enableMocks();

export const mockedNavigation = {
  navigate: jest.fn(),
  push: jest.fn(),
};

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");

  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigation.navigate,
      push: mockedNavigation.push,
    }),
  };
});

jest.mock('react-native-reanimated', () => {
  const Reanimated = require('react-native-reanimated/mock');

  // The mock for `call` immediately calls the callback which is incorrect
  // So we override it with a no-op
  Reanimated.default.call = () => {};

  return Reanimated;
});
