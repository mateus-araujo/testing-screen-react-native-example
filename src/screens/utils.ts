export enum TEST_IDS {
  SIGN_IN_USERNAME_INPUT = 'SignIn.usernameInput',
  SIGN_IN_PASSWORD_INPUT = 'SignIn.PasswordInput',
  SIGN_IN_BUTTON = 'SignIn.Button',
};
