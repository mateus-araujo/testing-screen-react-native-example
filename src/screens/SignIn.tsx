import { useNavigation } from "@react-navigation/native";
import React, { useState } from "react";
import { StyleSheet, Text, KeyboardAvoidingView } from "react-native";

import { Input, Button, ErrorText } from "../components/Form";
import { TEST_IDS } from "./utils"

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    justifyContent: "center",
    paddingHorizontal: 40,
  },
  headerText: {
    color: "#353031",
    fontWeight: "bold",
    fontSize: 34,
    marginBottom: 10,
  },
});

const useLoginFormState = ({ navigation }) => {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [submit, setSubmit] = useState(false);

  let isUsernameValid = false;
  let isPasswordValid = false;

  if (username === "example") {
    isUsernameValid = true;
  }

  if (password === "asdf") {
    isPasswordValid = true;
  }

  return {
    username: {
      value: username,
      set: setUsername,
      valid: isUsernameValid,
    },
    password: {
      value: password,
      set: setPassword,
      valid: isPasswordValid,
    },
    submit: {
      value: submit,
      set: setSubmit,
    },
  };
};

export default () => {
  const navigation = useNavigation();

  const { username, password, submit } = useLoginFormState({ navigation });

  let usernameErrorMsg;
  let passwordErrorMsg;

  if (submit.value && !username.valid) {
    usernameErrorMsg = "Invalid username.";
  }

  if (submit.value && !password.valid) {
    passwordErrorMsg = "Invalid password.";
  }

  const handleSubmit = () => {
    submit.set(true);

    if (username.valid && password.valid) {
      fetch("https://jsonplaceholder.typicode.com/users", {
        method: "POST",
        body: JSON.stringify({
          username,
          password,
        }),
      })
        .then(() => {
          navigation.navigate("App");
        })
        .catch((error) => {
          console.log("error", error);
        });
    }
  };

  return (
    <KeyboardAvoidingView style={styles.container} behavior="position">
      <Text style={styles.headerText}>Login</Text>
      <Input
        label="Username"
        placeholder="example"
        onChangeText={username.set}
        error={usernameErrorMsg}
        testID={TEST_IDS.SIGN_IN_USERNAME_INPUT}
      />
      <Input
        label="Password"
        placeholder="***"
        secureTextEntry
        onChangeText={password.set}
        error={passwordErrorMsg}
        testID={TEST_IDS.SIGN_IN_PASSWORD_INPUT}
      />
      <ErrorText messages={[usernameErrorMsg, passwordErrorMsg]} />
      <Button testID={TEST_IDS.SIGN_IN_BUTTON} text="Login" onPress={handleSubmit} />
    </KeyboardAvoidingView>
  );
};
