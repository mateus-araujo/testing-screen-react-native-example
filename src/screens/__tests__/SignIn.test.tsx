import React from "react";

import { act, fireEvent, render } from "@testing-library/react-native";
import fetch from "jest-fetch-mock"

import SignIn from "../SignIn";
import { TEST_IDS } from "../utils"
import { mockedNavigation } from "../../../jest/setup";


const flushMicrotasksQueue = () => new Promise((resolve) => setImmediate(resolve));

describe("SignIn", () => {  
  it("renders", () => {
    const { getAllByText, getByPlaceholderText } = render(<SignIn />);

    expect(getAllByText("Login").length).toBe(2);
    getByPlaceholderText("example");
    getByPlaceholderText("***");
  });

  it('shows invalid inputs messages', () => {
    const { getByTestId, getByText } = render(<SignIn />);

    fireEvent.press(getByTestId(TEST_IDS.SIGN_IN_BUTTON));

    getByText('Invalid username.')
    getByText('Invalid password.')
  });

  it('shows invalid user name error message', () => {
    const { getByTestId, getByText, queryAllByText } = render(<SignIn />);

    fireEvent.changeText(getByTestId(TEST_IDS.SIGN_IN_PASSWORD_INPUT), 'asdf');

    fireEvent.press(getByTestId(TEST_IDS.SIGN_IN_BUTTON));

    getByText('Invalid username.');
    expect(queryAllByText('Invalid password.')?.length).toBe(0);

    fireEvent.changeText(getByTestId(TEST_IDS.SIGN_IN_USERNAME_INPUT), 'invalid input');

    getByText('Invalid username.');
    expect(queryAllByText('Invalid password.')?.length).toBe(0);
  });

  it('shows invalid password. error message', () => {
    const { getByTestId, getByText, queryAllByText } = render(<SignIn />);

    fireEvent.changeText(getByTestId(TEST_IDS.SIGN_IN_USERNAME_INPUT), 'example');

    fireEvent.press(getByTestId(TEST_IDS.SIGN_IN_BUTTON));

    getByText('Invalid password.');
    expect(queryAllByText('Invalid username.')?.length).toBe(0);

    fireEvent.changeText(getByTestId(TEST_IDS.SIGN_IN_PASSWORD_INPUT), 'invalid input');

    getByText('Invalid password.');
    expect(queryAllByText('Invalid username.')?.length).toBe(0);
  });

  it('handles valid input submission', async () => {
    const { getByTestId } = render(<SignIn />);

    fireEvent.changeText(getByTestId(TEST_IDS.SIGN_IN_USERNAME_INPUT), 'example');
    fireEvent.changeText(getByTestId(TEST_IDS.SIGN_IN_PASSWORD_INPUT), 'asdf');
    fireEvent.press(getByTestId(TEST_IDS.SIGN_IN_BUTTON));

    expect(fetch.mock.calls).toMatchSnapshot();
    await act(flushMicrotasksQueue);

    expect(mockedNavigation.navigate).toBeCalledWith('App');
  });
});
