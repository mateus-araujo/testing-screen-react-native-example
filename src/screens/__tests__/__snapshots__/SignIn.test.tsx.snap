// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn handles valid input submission 1`] = `
Array [
  Array [
    "https://jsonplaceholder.typicode.com/users",
    Object {
      "body": "{\\"username\\":{\\"value\\":\\"example\\",\\"valid\\":true},\\"password\\":{\\"value\\":\\"asdf\\",\\"valid\\":true}}",
      "method": "POST",
    },
  ],
]
`;
